sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1986,66 +2015,27 @@ pi_result piDevicePartition(pi_device Device,
 
   PI_ASSERT(Device, PI_INVALID_DEVICE);
 
-  // Check if Device was already partitioned into the same or bigger size
-  // before. If so, we can return immediately without searching the global
-  // device cache. Note that L0 driver always returns the same handles in the
-  // same order for the given number of sub-devices.
-  if (OutDevices && NumDevices <= Device->SubDevices.size()) {
-    for (uint32_t I = 0; I < NumDevices; I++) {
-      OutDevices[I] = Device->SubDevices[I];
-      // reusing the same pi_device needs to increment the reference count
-      piDeviceRetain(OutDevices[I]);
-    }
-    if (OutNumDevices)
-      *OutNumDevices = NumDevices;
-    return PI_SUCCESS;
+  // Devices cache is normally created in piDevicesGet but still make
+  // sure that cache is populated.
+  //
+  pi_result Res = populateDeviceCacheIfNeeded(Device->Platform);
+  if (Res != PI_SUCCESS) {
+    return Res;
   }
 
-  // Get the number of subdevices available.
-  // TODO: maybe add interface to create the specified # of subdevices.
-  uint32_t Count = 0;
-  ZE_CALL(zeDeviceGetSubDevices, (Device->ZeDevice, &Count, nullptr));
-
   if (OutNumDevices) {
-    *OutNumDevices = Count;
+    *OutNumDevices = Device->SubDevices.size();
   }
 
-  if (!OutDevices) {
-    // If we are not given the buffer, we are done.
-    return PI_SUCCESS;
-  }
+  if (OutDevices) {
+    // TODO: Consider support for partitioning to <= total sub-devices.
+    PI_ASSERT(NumDevices == Device->SubDevices.size(), PI_INVALID_VALUE);
```

Comment:
I think we don't need this assert here.
We can just return the requested number of sub-devices (NumDevices).
Please note that the order of sub-device handles returned from L0 driver is always the same order.

Reply:
Today, the only supported partitioning doesn't have a way tp specify # of devices. So truly, SYCL RT will always ask for all sub-devices:
``` C++
        auto SubDevicesDomainNuma = rootdev.create_sub_devices<
            sycl::info::partition_property::partition_by_affinity_domain>(
            sycl::info::partition_affinity_domain::numa);
```

Reply:
I just wanted to check this assumption explicitly.

Reply:
I thought we support partitioning of sub-devices, which means partitioning multiple levels.
In Sycl spec 2020 page 64,
<pre>
A SYCL device can be partitioned into multiple SYCL devices, by calling the create_sub_devices() member
function template. The resulting SYCL devices are considered sub devices, and it is valid to partition
these sub devices further.
</pre>

Reply:
One should first query the supported partitioning, and Level-Zero would only return the one I showed.

Reply:
Do you mean partition_by_affinity does not support partitioning multiple levels?
If so, could you add a comment about this fact for other people?
"We currently support only partition_by_affinity which does not allow partitioning multiple levels. So we always return the maximum number of sub-devices."
That will be very useful to understand the logic behind this code.

Reply:
what is "multiple levels" in your statement?

Reply:
What I mean by multiple levels is calling partition_device() on the sub-devices which are already partitioned before.

Reply:
> partition these sub devices further.

That's not [yet] supported. I will add a comment about why we always expect all devices requested.